File: src/main/java/br/com/correios/enderecador/telas/TelaPesquisarDestinatario.java
package br.com.correios.enderecador.telas;

import br.com.correios.enderecador.bean.DestinatarioBean;
import br.com.correios.enderecador.dao.DaoException;
import br.com.correios.enderecador.dao.DestinatarioDao;
import br.com.correios.enderecador.util.Geral;

import java.awt.Dimension;
import java.awt.Font;
import java.awt.Frame;
import java.awt.LayoutManager;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Vector;
import javax.swing.BorderFactory;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextField;
import javax.swing.JToolBar;
import javax.swing.ListModel;

import org.apache.log4j.Logger;
import org.netbeans.lib.awtextra.AbsoluteConstraints;
import org.netbeans.lib.awtextra.AbsoluteLayout;

public class TelaPesquisarDestinatario extends JDialog {
    private static final Logger logger = Logger.getLogger(TelaPesquisarDestinatario.class);

    private final Vector<DestinatarioBean> vecDestinatario;

    private Vector<DestinatarioBean> vecDestinatarioRetorno;

    private DestinatarioBean destinatarioBean;

    private String ultimaConsulta;

    private JScrollPane jScrollPane1;

    private JList jlstDestinatario;

    private JTextField jtxtDestinatario;

    public TelaPesquisarDestinatario(Frame parent, boolean modal, Vector<DestinatarioBean> vecDestinatario) {
        super(parent, modal);
        this.vecDestinatarioRetorno = new Vector<>();
        this.vecDestinatario = new Vector<>();
        this.ultimaConsulta = "";
        initComponents();
        this.vecDestinatarioRetorno = vecDestinatario;
        carregaListaDestinatario();
    }

    private void carregaListaDestinatario() {
        try {
            ArrayList<DestinatarioBean> arrayDestinatario = DestinatarioDao.getInstance().recuperaDestinatario("");
            this.vecDestinatario.addAll(arrayDestinatario);
            this.jlstDestinatario.setListData(this.vecDestinatario);
        } catch (DaoException e) {
            logger.error(e.getMessage(), e);
            JOptionPane.showMessageDialog(this, "Não foi possivel carregar relação de destinatários", "Endereçador ", JOptionPane.WARNING_MESSAGE);
        }
    }

    private void initComponents() {
        JToolBar jToolBar1 = new JToolBar();
        JButton jbtConfirmar = new JButton();
        JButton jbtPesquisar = new JButton();
        JButton jbtSair = new JButton();
        JPanel jPanel1 = new JPanel();
        JLabel jLabel1 = new JLabel();
        this.jtxtDestinatario = new JTextField();
        this.jScrollPane1 = new JScrollPane();
        this.jlstDestinatario = new JList<>();
        setDefaultCloseOperation(2);
        setTitle("Selecionar destinatário");
        setResizable(false);
        jToolBar1.setBorder(BorderFactory.createEtchedBorder());
        jbtConfirmar.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 9));
        jbtConfirmar.setIcon(new ImageIcon(getClass().getResource("/imagens/OK.gif")));
        jbtConfirmar.setText("Confirmar");
        jbtConfirmar.setHorizontalTextPosition(0);
        jbtConfirmar.setMaximumSize(new Dimension(90, 60));
        jbtConfirmar.setVerticalTextPosition(3);
        jbtConfirmar.addActionListener(TelaPesquisarDestinatario.this::jbtConfirmarActionPerformed);
        jToolBar1.add(jbtConfirmar);
        jbtPesquisar.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 9));
        jbtPesquisar.setIcon(new ImageIcon(getClass().getResource("/imagens/binoculo.gif")));
        jbtPesquisar.setText("Pesquisar");
        jbtPesquisar.setHorizontalTextPosition(0);
        jbtPesquisar.setMaximumSize(new Dimension(90, 60));
        jbtPesquisar.setVerticalTextPosition(3);
        jbtPesquisar.addActionListener(TelaPesquisarDestinatario.this::jbtPesquisarActionPerformed);
        jToolBar1.add(jbtPesquisar);
        jbtSair.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 9));
        jbtSair.setIcon(new ImageIcon(getClass().getResource("/imagens/sair.gif")));
        jbtSair.setText("Sair");
        jbtSair.setHorizontalTextPosition(0);
        jbtSair.setMaximumSize(new Dimension(90, 60));
        jbtSair.setVerticalTextPosition(3);
        jbtSair.addActionListener(TelaPesquisarDestinatario.this::jbtSairActionPerformed);
        jToolBar1.add(jbtSair);
        getContentPane().add(jToolBar1, "North");
        jPanel1.setLayout(new AbsoluteLayout());
        jPanel1.setBorder(BorderFactory.createEtchedBorder());
        jLabel1.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 10));
        jLabel1.setText("Destinatário:");
        jPanel1.add(jLabel1, new AbsoluteConstraints(20, 23, -1, -1));
        jPanel1.add(this.jtxtDestinatario, new AbsoluteConstraints(90, 20, 350, -1));
        this.jScrollPane1.setViewportView(this.jlstDestinatario);
        jPanel1.add(this.jScrollPane1, new AbsoluteConstraints(4, 52, 582, 210));
        getContentPane().add(jPanel1, "Center");
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        setBounds((screenSize.width - 598) / 2, (screenSize.height - 349) / 2, 598, 349);
    }

    private void jbtSairActionPerformed(ActionEvent evt) {
        setVisible(false);
    }

    private void jbtPesquisarActionPerformed(ActionEvent evt) {
        try {
            String nomeDestinatario = this.jtxtDestinatario.getText().trim();
            int index = -1;
            if (this.ultimaConsulta.equalsIgnoreCase(nomeDestinatario)) {
                index = this.jlstDestinatario.getSelectedIndex();
            } else {
                this.ultimaConsulta = nomeDestinatario;
            }
            for (int i = index + 1; i < this.vecDestinatario.size(); i++) {
                this.destinatarioBean = this.vecDestinatario.get(i);
                String nomeDestinatarioTabela = this.destinatarioBean.getNome();
                if (nomeDestinatarioTabela.toUpperCase().contains(nomeDestinatario.toUpperCase())) {
                    index = i;
                    break;
                }
            }
            if (index == -1) {
                JOptionPane.showMessageDialog(this, "Destinatário não encontrado!", "Endereçador", JOptionPane.INFORMATION_MESSAGE);
            } else {
                this.jlstDestinatario.setSelectedIndex(index);
                int tamanhoJscrool = this.jScrollPane1.getVerticalScrollBar().getMaximum();
                int value = 0;
                value = index * tamanhoJscrool / this.vecDestinatario.size();
                this.jScrollPane1.getVerticalScrollBar().setValue(value);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private void jbtConfirmarActionPerformed(ActionEvent evt) {
        if (this.jlstDestinatario.isSelectionEmpty()) {
            JOptionPane.showMessageDialog(this, "Não existe nenhum destinatário selecionado", "Endereçador ECT", 1);
            return;
        }
        ListModel<DestinatarioBean> modelDestinatario = this.jlstDestinatario.getModel();
        Geral geral = null;
        int[] arrayIndex = this.jlstDestinatario.getSelectedIndices();
        this.destinatarioBean = DestinatarioBean.getInstance();
        for (int index : arrayIndex) {
            this.destinatarioBean = modelDestinatario.getElementAt(index);
            if (!Geral.verificaExistencia(this.destinatarioBean, this.vecDestinatarioRetorno))
                this.vecDestinatarioRetorno.add(this.destinatarioBean);
        }
        Geral.ordenarVetor(this.vecDestinatarioRetorno, this.destinatarioBean);
        setVisible(false);
    }
}

File: src/main/kotlin/br/com/correios/enderecador/telas/TelaPesquisarDestinatario.kt
package br.com.correios.enderecador.telas

import br.com.correios.enderecador.bean.DestinatarioBean
import javax.swing.JDialog
import javax.swing.JScrollPane
import javax.swing.JList
import javax.swing.JTextField
import br.com.correios.enderecador.dao.DestinatarioDao
import br.com.correios.enderecador.dao.DaoException
import javax.swing.JOptionPane
import javax.swing.JToolBar
import javax.swing.JButton
import javax.swing.JPanel
import javax.swing.JLabel
import javax.swing.BorderFactory
import javax.swing.ImageIcon
import java.awt.event.ActionEvent
import org.netbeans.lib.awtextra.AbsoluteLayout
import org.netbeans.lib.awtextra.AbsoluteConstraints
import javax.swing.ListModel
import br.com.correios.enderecador.util.Geral
import org.apache.log4j.Logger
import java.awt.Dimension
import java.awt.Font
import java.awt.Frame
import java.awt.Toolkit
import java.lang.Exception
import java.util.*

class TelaPesquisarDestinatario(parent: Frame?, modal: Boolean, vecDestinatario: Vector<DestinatarioBean?>) :
    JDialog(parent, modal) {
    private val vecDestinatario: Vector<DestinatarioBean>
    private var vecDestinatarioRetorno: Vector<DestinatarioBean?>
    private var destinatarioBean: DestinatarioBean? = null
    private var ultimaConsulta: String
    private var jScrollPane1: JScrollPane? = null
    private var jlstDestinatario: JList<DestinatarioBean>? = null
    private var jtxtDestinatario: JTextField? = null

    init {
        vecDestinatarioRetorno = Vector()
        this.vecDestinatario = Vector()
        ultimaConsulta = ""
        initComponents()
        vecDestinatarioRetorno = vecDestinatario
        carregaListaDestinatario()
    }

    private fun carregaListaDestinatario() {
        try {
            val arrayDestinatario = DestinatarioDao.instance!!.recuperaDestinatario("")
            vecDestinatario.addAll(arrayDestinatario)
            jlstDestinatario!!.setListData(vecDestinatario)
        } catch (e: DaoException) {
            logger.error(e.message, e)
            JOptionPane.showMessageDialog(
                this,
                "Não foi possivel carregar relação de destinatários",
                "Endereçador ",
                JOptionPane.WARNING_MESSAGE
            )
        }
    }

    private fun initComponents() {
        val jToolBar1 = JToolBar()
        val jbtConfirmar = JButton()
        val jbtPesquisar = JButton()
        val jbtSair = JButton()
        val jPanel1 = JPanel()
        val jLabel1 = JLabel()
        jtxtDestinatario = JTextField()
        jScrollPane1 = JScrollPane()
        jlstDestinatario = JList<DestinatarioBean>()
        defaultCloseOperation = 2
        title = "Selecionar destinatário"
        isResizable = false
        jToolBar1.border = BorderFactory.createEtchedBorder()
        jbtConfirmar.font = Font(Font.SANS_SERIF, Font.PLAIN, 9)
        jbtConfirmar.icon = ImageIcon(javaClass.getResource("/imagens/OK.gif"))
        jbtConfirmar.text = "Confirmar"
        jbtConfirmar.horizontalTextPosition = 0
        jbtConfirmar.maximumSize = Dimension(90, 60)
        jbtConfirmar.verticalTextPosition = 3
        jbtConfirmar.addActionListener { evt: ActionEvent -> jbtConfirmarActionPerformed(evt) }
        jToolBar1.add(jbtConfirmar)
        jbtPesquisar.font = Font(Font.SANS_SERIF, Font.PLAIN, 9)
        jbtPesquisar.icon = ImageIcon(javaClass.getResource("/imagens/binoculo.gif"))
        jbtPesquisar.text = "Pesquisar"
        jbtPesquisar.horizontalTextPosition = 0
        jbtPesquisar.maximumSize = Dimension(90, 60)
        jbtPesquisar.verticalTextPosition = 3
        jbtPesquisar.addActionListener { evt: ActionEvent -> jbtPesquisarActionPerformed(evt) }
        jToolBar1.add(jbtPesquisar)
        jbtSair.font = Font(Font.SANS_SERIF, Font.PLAIN, 9)
        jbtSair.icon = ImageIcon(javaClass.getResource("/imagens/sair.gif"))
        jbtSair.text = "Sair"
        jbtSair.horizontalTextPosition = 0
        jbtSair.maximumSize = Dimension(90, 60)
        jbtSair.verticalTextPosition = 3
        jbtSair.addActionListener { evt: ActionEvent -> jbtSairActionPerformed(evt) }
        jToolBar1.add(jbtSair)
        contentPane.add(jToolBar1, "North")
        jPanel1.layout = AbsoluteLayout()
        jPanel1.border = BorderFactory.createEtchedBorder()
        jLabel1.font = Font(Font.SANS_SERIF, Font.PLAIN, 10)
        jLabel1.text = "Destinatário:"
        jPanel1.add(jLabel1, AbsoluteConstraints(20, 23, -1, -1))
        jPanel1.add(jtxtDestinatario, AbsoluteConstraints(90, 20, 350, -1))
        jScrollPane1!!.setViewportView(jlstDestinatario)
        jPanel1.add(jScrollPane1, AbsoluteConstraints(4, 52, 582, 210))
        contentPane.add(jPanel1, "Center")
        val screenSize = Toolkit.getDefaultToolkit().screenSize
        setBounds((screenSize.width - 598) / 2, (screenSize.height - 349) / 2, 598, 349)
    }

    private fun jbtSairActionPerformed(evt: ActionEvent) {
        isVisible = false
    }

    private fun jbtPesquisarActionPerformed(evt: ActionEvent) {
        try {
            val nomeDestinatario = jtxtDestinatario!!.text.trim { it <= ' ' }
            var index = -1
            if (ultimaConsulta.equals(nomeDestinatario, ignoreCase = true)) {
                index = jlstDestinatario!!.selectedIndex
            } else {
                ultimaConsulta = nomeDestinatario
            }
            for (i in index + 1 until vecDestinatario.size) {
                destinatarioBean = vecDestinatario[i]
                val nomeDestinatarioTabela = destinatarioBean!!.nome
                if (nomeDestinatarioTabela.uppercase(Locale.getDefault())
                        .contains(nomeDestinatario.uppercase(Locale.getDefault()))
                ) {
                    index = i
                    break
                }
            }
            if (index == -1) {
                JOptionPane.showMessageDialog(
                    this,
                    "Destinatário não encontrado!",
                    "Endereçador",
                    JOptionPane.INFORMATION_MESSAGE
                )
            } else {
                jlstDestinatario!!.selectedIndex = index
                val tamanhoJscrool = jScrollPane1!!.verticalScrollBar.maximum
                var value = 0
                value = index * tamanhoJscrool / vecDestinatario.size
                jScrollPane1!!.verticalScrollBar.value = value
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    private fun jbtConfirmarActionPerformed(evt: ActionEvent) {
        if (jlstDestinatario!!.isSelectionEmpty) {
            JOptionPane.showMessageDialog(this, "Não existe nenhum destinatário selecionado", "Endereçador ECT", 1)
            return
        }
        val modelDestinatario: ListModel<DestinatarioBean> = jlstDestinatario!!.getModel()
        val arrayIndex = jlstDestinatario!!.selectedIndices
        destinatarioBean = DestinatarioBean.instance
        for (index in arrayIndex) {
            destinatarioBean = modelDestinatario.getElementAt(index)
            if (!Geral.verificaExistencia(destinatarioBean, vecDestinatarioRetorno)) {
                vecDestinatarioRetorno.add(destinatarioBean)
            }
        }
        Geral.ordenarVetor(vecDestinatarioRetorno, destinatarioBean)
        isVisible = false
    }

    companion object {
        private val logger = Logger.getLogger(TelaPesquisarDestinatario::class.java)
    }
}
